backend/xivdpscalc/pps/sch.py

```diff
@@ -12,6 +17,71 @@ class SchPps(HealerPps):
     bio_potency = 70
     ed_potency = 100
 
+    def get_total_potency_variable_time(self, sim_length: ElapsedTime, character_stats: Character, rotation: SchRotation, caster_tax: float = 0.1) -> float:
+        """
+        Get the total potency in sim_length seconds
+        rotation is a SchRotation object
+        todo: support dot damage calculation
+        todo: how long is animation lock actually?  Hard coded here for 0.75, have no idea if that's true
+        There are some nuances to how this works:
+            At present, this sim does not calculate dot damage
+            If an action with an active cooldown is selected, the time jumps forward to the cooldown's expiry
+            This simulator does not enforce resource requirements - ED can be used even with no aetherflow
+        """
+        animation_lock = 0.75
+
+        total_potency = 0
+        current_time: ElapsedTime = 0
+
+        short_gcd = character_stats.get_gcd()
+
+        next_active: defaultdict[SchAction, ElapsedTime] = defaultdict(lambda: 0)
+        active_effects: defaultdict[SchEffect, ElapsedTime] = defaultdict(lambda: -1)
+        resources: defaultdict[SchResource, int] = defaultdict(lambda: 0)
+
+        while current_time <= sim_length:
+            selected_action = rotation.get_action(current_time,
+                                                  {key: next_active[key] - current_time for key in SchAction},
+                                                  {key: active_effects[key] - current_time for key in active_effects},
+                                                  resources)
+
+
+            cast_time = character_stats.get_cast_time(selected_action.cast_time)
+
+            if active_effects[SchEffect.SWIFTCAST] > current_time and cast_time > 0:
+                active_effects[SchEffect.SWIFTCAST] = 0
+                cast_time = 0
+
+            if current_time + cast_time <= sim_length \
+                and not selected_action.is_dot:
+                total_potency += selected_action.potency
+
+            current_time = max(current_time, next_active[selected_action])
+            next_active[selected_action] = current_time + selected_action.cooldown
+
+            if selected_action.is_gcd:
+                for action in SchAction:
+                    if action.is_gcd:
+                        next_active[action] = current_time + short_gcd
+
+            current_time += \
+                max(cast_time + caster_tax, animation_lock)
+
+            if selected_action in [SchAction.AETHERFLOW, SchAction.DISSIPATION]:
+                resources[SchResource.AETHERFLOW] = 3
+            elif selected_action == SchAction.ENERGYDRAIN:
```

Comment:
there should be some kind of check on this one. please don't energy drain yourself into negative stacks.

Reply:
actually the more i thought about this, the more i think we should assume the rotation is valid.

which means that there should be a preprocessing step in a separate function that merely checks for validity.

reason being this rotation object will eventually get passed around to a bunch of stuff (non-dot damage, dot damage, mp sim) and checking validity repeatedly sounds like a pain in the ass.